zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,726 @@
+	list_item_append(&vmh_list, &new_heap->node);
+
+	return new_heap;
+fail:
+	for (i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		struct sys_mem_blocks *allocator = new_heap->physical_blocks_allocators[i];
+		struct sys_bitarray *allocation_sizes = new_heap->allocation_sizes[i];
+
+		if (allocator) {
+			if (allocator->bitmap)
+				rfree(allocator->bitmap->bundles);
+			rfree(allocator->bitmap);
+		}
+		if (allocation_sizes) {
+			if (allocation_sizes->bundles)
+				rfree(allocation_sizes->bundles);
+			rfree(allocation_sizes);
+		}
+		rfree(allocator);
+	}
+	rfree(new_heap);
+	return NULL;
+}
+
+/**
+ * @brief Alloc function
+ *
+ * Allocates memory on heap from provided heap pointer.
+ * Check if we need to map physical memory for given allocation
+ * and do it when necessary.
+ * @param heap pointer to a heap that will be used for allocation.
+ * @param alloc_size Size in bytes to allocate.
+ *
+ * @retval ptr to allocated memory if successful
+ * @retval NULL on allocation failure
+ */
+void *vmh_alloc(struct virtual_memory_heap *heap, uint32_t alloc_size)
+{
+	if (!alloc_size)
+		return NULL;
+
+	void *ptr = NULL;
+	void *phys_aligned_ptr, *phys_aligned_alloc_end;
+	uint32_t block_size, allocation_error_code;
+	size_t i, mem_block_iterator, allocation_bitarray_offset, allocation_bitarray_position,
+		check_offset, check_size, check_position, block_count, physical_block_count;
+	uintptr_t phys_block_ptr, phys_block_end, block_realigned_ptr,
+		block_realigned_alloc_end;
+
+	/* We will gather error code when allocating on physical block
+	 * allocators.
+	 * Loop will continue as long as there are blocks to iterate on and
+	 * if allocation keeps returning error code.
+	 * When alloc succeeded it will return 0 on allocation and we break
+	 * form loop and continue.
+	 * If span is on for heap, allocation will be made on first available
+	 * block series on given block bundle. For span off allocation will only
+	 * happened on first free block on first allocator that can accommodate
+	 * requested size.
+	 */
+	for (mem_block_iterator = 0; mem_block_iterator < MAX_MEMORY_ALLOCATORS_COUNT;
+	     mem_block_iterator++) {
+		/* calculate block count needed to allocate for current
+		 * mem_block.
+		 */
+		block_size =
+		    1 << heap->physical_blocks_allocators[mem_block_iterator]->blk_sz_shift;
+		block_count = SOF_DIV_ROUND_UP(alloc_size, block_size);
+
+		if (block_count >
+		    heap->physical_blocks_allocators[mem_block_iterator]->num_blocks)
+			continue;
+		/* Try span alloc on first available mem_block for non span
+		 * check if block size is sufficient.
+		 */
+		if (heap->span) {
+			allocation_error_code = sys_mem_blocks_alloc_contiguous(
+				heap->physical_blocks_allocators[mem_block_iterator], block_count,
+				&ptr);
+		} else if (block_size > alloc_size) {
+			allocation_error_code = sys_mem_blocks_alloc(
+				heap->physical_blocks_allocators[mem_block_iterator], block_count,
+				&ptr);
+		}
```

Comment:
I see no option for it since this is always a span or non span allocation exclusively.
Can You specify the scenario ?
We break out of the loop whenever allocation was successfully made so i can see no issue there too.

Reply:
got ya